prov/sm2/src/sm2_progress.c

```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+					 (void *) hmem_addr, bytes_to_send);
+		payload_used += bytes_to_send;
+		ofi_consume_rma_iov(cmd_rma->rma_iov, &rma_count,
+				    bytes_to_send);
+
+		assert(payload_used <= SM2_RMA_INJECT_SIZE);
+		if (ret) {
+			cmd_rma->sar_hdr.proto_flags |= FI_SM2_SAR_ERROR_FLAG;
+			break;
+		}
+	}
+	xfer_entry->hdr.op = ofi_op_read_rsp;
+	sm2_fifo_write(ep, xfer_entry->hdr.sender_gid, xfer_entry);
+	return FI_SUCCESS;
+}
+
+static int sm2_progress_rma_read_rsp(struct sm2_ep *ep,
+				     struct sm2_xfer_entry *xfer_entry)
+{
+	int ret = 0;
+	struct sm2_cmd_sar_rma_msg *cmd_rma = (void *) xfer_entry->user_data;
+	struct sm2_sar_ctx *ctx = (void *) cmd_rma->sar_hdr.context;
+	struct iovec iov[SM2_IOV_LIMIT];
+	void *desc[SM2_IOV_LIMIT];
+	enum fi_hmem_iface iface;
+	uint64_t device;
+	size_t bytes_to_copy, iov_count, bytes_used = 0;
+
+	assert(ctx);
+	ctx->msgs_in_flight--;
+
+	cmd_rma = (struct sm2_cmd_sar_rma_msg *) xfer_entry->user_data;
+	memcpy(iov, ctx->msg.msg_iov, sizeof(iov));
+	memcpy(desc, ctx->msg.desc, sizeof(desc));
+	iov_count = SM2_IOV_LIMIT;
+	if (cmd_rma->sar_hdr.request_offset) {
+		ofi_consume_iov_desc(iov, desc, &iov_count,
+				     cmd_rma->sar_hdr.request_offset);
+	}
+
+	ctx->status_flags |=
+		cmd_rma->sar_hdr.proto_flags & FI_SM2_SAR_ERROR_FLAG;
+	if (ctx->status_flags & FI_SM2_SAR_ERROR_FLAG) {
+		sm2_rma_handle_remote_error(ep, xfer_entry, ctx);
+		return FI_SUCCESS;
+	}
+
+	while (iov_count) {
+		bytes_to_copy =
+			MIN(iov[0].iov_len, SM2_RMA_INJECT_SIZE - bytes_used);
+		if (bytes_to_copy == 0)
+			break;
+		sm2_get_iface_device((void **) desc, &iface, &device);
+		ret = ofi_copy_to_hmem(iface, device, iov[0].iov_base,
+				       cmd_rma->user_data + bytes_used,
+				       bytes_to_copy);
+		bytes_used += bytes_to_copy;
+		assert(bytes_used <= SM2_RMA_INJECT_SIZE);
+		ofi_consume_iov_desc(iov, desc, &iov_count, bytes_to_copy);
+		if (ret)
+			break;
+	}
+	ctx->bytes_acked += bytes_used;
+
+	/* report completion if this is the last packet. */
+	if (ctx->bytes_acked == ctx->bytes_total) {
+		uint64_t comp_flags = ofi_tx_cq_flags(xfer_entry->hdr.op);
+		ret = sm2_complete_tx(ep, (void *) xfer_entry->hdr.context,
+				      xfer_entry->hdr.op,
+				      comp_flags | FI_COMPLETION);
+		sm2_free_rma_ctx(ctx);
+		sm2_freestack_push(ep, xfer_entry);
+		return FI_SUCCESS;
+	}
+
+	if (ctx->bytes_requested == ctx->bytes_total) {
+		sm2_freestack_push(ep, xfer_entry);
+		return FI_SUCCESS;
+	}
+
+	ret = sm2_rma_cmd_fill_sar_xfer(xfer_entry, ctx);
+	if (!ret)
+		sm2_fifo_write(ep, ctx->peer_gid, xfer_entry);
+	else
+		sm2_rma_handle_local_error(ep, xfer_entry, ctx, ret);
+
+	return FI_SUCCESS;
+}
+
+static int sm2_progress_rma_write_origin(struct sm2_ep *ep,
+					 struct sm2_xfer_entry *xfer_entry)
+{
+	struct sm2_cmd_sar_rma_msg *cmd_rma = (void *) xfer_entry->user_data;
+	struct sm2_sar_ctx *ctx = (void *) cmd_rma->sar_hdr.context;
+	bool tx_complete, rx_complete;
+	int ret;
+
+	if (!ctx) {
```

Comment:
It looks to me like rma_inject useses generic_rma which creates a context for every message... so this branch is dead code which will never be hit.

Reply:
Thank you, old code :-)